File: apps/kcoder-studio/renderer/packages/chat-core/src/provider-failure.ts
const categories = [
  'authentication_error',
  'forbidden',
  'model_or_route',
  'invalid_parameter',
  'context_length_exceeded',
  'quota_exceeded',
  'rate_limit',
  'network_error',
  'timeout_error',
  'model_protocol_error',
  'provider_error',
] as const

export interface ProviderFailureDetails {
  category: (typeof categories)[number]
  recovery_action: 'needs_human' | 'diagnose_only'
  http_status?: number | null
  retryable: boolean
  resume_safe: boolean
  retry_after_ms?: number | null
}

export function decodeProviderFailure(value: unknown): ProviderFailureDetails | undefined {
  if (!value || typeof value !== 'object' || Array.isArray(value)) return undefined
  const data = value as Record<string, unknown>
  if (
    !categories.includes(data.category as ProviderFailureDetails['category']) ||
    !['needs_human', 'diagnose_only'].includes(data.recovery_action as string) ||
    typeof data.retryable !== 'boolean' ||
    typeof data.resume_safe !== 'boolean'
  )
    return undefined
  if (
    (data.resume_safe && !data.retryable) ||
    (data.recovery_action === 'needs_human' && (data.retryable || data.resume_safe))
  )
    return undefined
  for (const [key, min, max] of [
    ['http_status', 100, 999],
    ['retry_after_ms', 0, Number.MAX_SAFE_INTEGER],
  ] as const) {
    const number = data[key]
    if (
      number !== undefined &&
      number !== null &&
      (typeof number !== 'number' || !Number.isSafeInteger(number) || number < min || number > max)
    )
      return undefined
  }
  return {
    category: data.category as ProviderFailureDetails['category'],
    recovery_action: data.recovery_action as ProviderFailureDetails['recovery_action'],
    retryable: data.retryable,
    resume_safe: data.resume_safe,
    ...(data.http_status !== undefined ? { http_status: data.http_status as number | null } : {}),
    ...(data.retry_after_ms !== undefined
      ? { retry_after_ms: data.retry_after_ms as number | null }
      : {}),
  }
}

File: crates/kcoder_types/src/provider_failure.rs
use serde::{Deserialize, Serialize};

/// Allowlisted terminal facts, independent of the provider's diagnostic text.
///
/// Three different questions are answered here, and a consumer must not collapse
/// them into one "can I retry?" flag:
///
/// - [`Self::retryable`] — repeating the *request* is allowed. It says nothing
///   about whether the failed attempt already produced visible output.
/// - [`Self::resume_safe`] — repeating it cannot duplicate a response that had
///   already started streaming, so the conversation can be resumed as it stands.
/// - [`Self::recovery_action`] — whether a human has to act at all.
///
/// None of them means "this failed turn can be continued from its committed
/// boundary": that is a separate server capability
/// (`kcoder_app_protocol::CAPABILITY_FAILED_TURN_CONTINUATION_V1`, requested with
/// `turn/start.retryFromTurnId`), and a client must gate on the capability rather
/// than infer it from these booleans.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct ProviderFailureDetails {
    pub category: ProviderFailureCategory,
    pub recovery_action: ProviderFailureRecoveryAction,
    pub http_status: Option<u16>,
    pub retryable: bool,
    pub resume_safe: bool,
    pub retry_after_ms: Option<u64>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProviderFailureCategory {
    AuthenticationError,
    Forbidden,
    ModelOrRoute,
    InvalidParameter,
    ContextLengthExceeded,
    RateLimit,
    QuotaExceeded,
    NetworkError,
    TimeoutError,
    ModelProtocolError,
    ProviderError,
}

impl ProviderFailureCategory {
    pub fn as_str(self) -> &'static str {
        match self {
            Self::AuthenticationError => "authentication_error",
            Self::Forbidden => "forbidden",
            Self::ModelOrRoute => "model_or_route",
            Self::InvalidParameter => "invalid_parameter",
            Self::ContextLengthExceeded => "context_length_exceeded",
            Self::RateLimit => "rate_limit",
            Self::QuotaExceeded => "quota_exceeded",
            Self::NetworkError => "network_error",
            Self::TimeoutError => "timeout_error",
            Self::ModelProtocolError => "model_protocol_error",
            Self::ProviderError => "provider_error",
        }
    }
}

/// Terminal instructions never schedule another invocation or approval request.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ProviderFailureRecoveryAction {
    NeedsHuman,
    DiagnoseOnly,
}
